pyomo/contrib/solver/config.py

```diff
@@ -0,0 +1,377 @@
+#  ___________________________________________________________________________
+
+from typing import Optional
+
+from pyomo.common.config import (
+    ConfigDict,
+    ConfigValue,
+    NonNegativeFloat,
+    NonNegativeInt,
+    ADVANCED_OPTION,
+)
+from pyomo.common.timing import HierarchicalTimer
+
+
+class SolverConfig(ConfigDict):
+    """
+    Base config for all direct solver interfaces
+    """
+
+    def __init__(
+        self,
+        description=None,
+        doc=None,
+        implicit=False,
+        implicit_domain=None,
+        visibility=0,
+    ):
+        super().__init__(
+            description=description,
+            doc=doc,
+            implicit=implicit,
+            implicit_domain=implicit_domain,
+            visibility=visibility,
+        )
+
+        self.tee: bool = self.declare(
+            'tee',
+            ConfigValue(
+                domain=bool,
+                default=False,
+                description="If True, the solver log prints to stdout.",
+            ),
+        )
+        self.log_solver_output: bool = self.declare(
+            'log_solver_output',
+            ConfigValue(
+                domain=bool,
+                default=False,
+                description="If True, the solver output gets logged.",
+            ),
+        )
+        self.working_dir: str = self.declare(
+            'working_dir',
+            ConfigValue(
+                domain=str,
+                default=None,
+                description="The directory in which generated files should be saved. "
+                "This replaces the `keepfiles` option.",
+            ),
+        )
+        self.load_solutions: bool = self.declare(
+            'load_solutions',
+            ConfigValue(
+                domain=bool,
+                default=True,
+                description="If True, the values of the primal variables will be loaded into the model.",
+            ),
+        )
+        self.raise_exception_on_nonoptimal_result: bool = self.declare(
+            'raise_exception_on_nonoptimal_result',
+            ConfigValue(
+                domain=bool,
+                default=True,
+                description="If False, the `solve` method will continue processing "
+                "even if the returned result is nonoptimal.",
+            ),
+        )
+        self.symbolic_solver_labels: bool = self.declare(
+            'symbolic_solver_labels',
+            ConfigValue(
+                domain=bool,
+                default=False,
+                description="If True, the names given to the solver will reflect the names of the Pyomo components. "
+                "Cannot be changed after set_instance is called.",
+            ),
+        )
+        self.timer: HierarchicalTimer = self.declare(
+            'timer', ConfigValue(default=None, description="A HierarchicalTimer.")
+        )
+        self.threads: Optional[int] = self.declare(
```

Comment:
@michaelbynum - this is a you question. I'm not sure if using `Optional` actually benefits us unless it's something relating to typing?

Reply:
I updated the type hints to reflect the type we expect. For all of the `bool` options, we are *not* okay with `None`, so I left those as `bool` instead of `Optional[bool]`.

Reply:
Unfortunately, this is a limitation of the current Config system: None is *always* admissible.  But I agree that documenting that we don't want None is a good thing.